ingestor-default/src/main/scala/za/co/absa/hyperdrive/ingestor/implementation/reader/parquet/ParquetStreamReader.scala

```diff
@@ -0,0 +1,96 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.hyperdrive.ingestor.implementation.reader.parquet
+
+import java.util.concurrent.CountDownLatch
+
+import org.apache.commons.configuration2.Configuration
+import org.apache.commons.lang3.StringUtils
+import org.apache.logging.log4j.LogManager
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import za.co.absa.hyperdrive.ingestor.api.reader.{StreamReader, StreamReaderFactory}
+import za.co.absa.hyperdrive.ingestor.api.utils.{ComponentFactoryUtil, ConfigUtils}
+import za.co.absa.hyperdrive.shared.utils.FileUtils
+
+import scala.util.Try
+
+/**
+ * Creates a parquet reader for a file source.
+ *
+ * @param path       Path to the file source.
+ * @param extraConfs Extra configurations, see Spark documentation
+ */
+private[reader] class ParquetStreamReader(
+  val path: String,
+  val waitForFiles: Boolean,
+  val checkForFilesInterval: Long,
+  val extraConfs: Map[String, String]) extends StreamReader {
+
+  private val logger = LogManager.getLogger()
+  private val waitingForFilesLatch = new CountDownLatch(1)
+
+  if (StringUtils.isBlank(path)) {
+    throw new IllegalArgumentException(s"No path specified")
+  }
+
+  def awaitWaitingForFiles(): Unit = waitingForFilesLatch.await()
+
+  override def read(spark: SparkSession): DataFrame = {
+
+    if (spark.sparkContext.isStopped) {
+      throw new IllegalStateException("SparkSession is stopped.")
+    }
+
+    logger.info(s"Will read from path $path")
+
+    if(waitForFiles) {
+      waitingForFilesLatch.countDown()
+      val fs = spark.sparkContext.hadoopConfiguration
+      while (FileUtils.dirContainsNoParquetFilesOrDoesNotExist(path, fs)) {
+        Thread.sleep(checkForFilesInterval)
+      }
+    }
+
+    val schema = spark.read.load(path).schema
+
+    spark
+      .readStream
+      .schema(schema)
+      .options(extraConfs)
+      .parquet(path)
+  }
+}
+
+object ParquetStreamReader extends StreamReaderFactory with ParquetStreamReaderAttributes {
+  private val logger = LogManager.getLogger
+  private val defaultCheckForInitialFileInterval = "10000"
+
+  override def apply(conf: Configuration): StreamReader = {
+    ComponentFactoryUtil.validateConfiguration(conf, getProperties)
+
+    val path = conf.getString(KeySourceDirectory)
+    val waitForFiles = Try(conf.getString(KeyWaitForFiles)).toOption.isDefined
+    val checkForInitialFileInterval = Try(conf.getString(KeyCheckForInitialFileInterval))
+      .getOrElse(defaultCheckForInitialFileInterval)
+      .toLong
```

Comment:
It may be more concise to use the `getString(key, default)` method directly

Reply:
This implementation was broken. The `getOrElse` was on the `Try` which makes no sense. After another thought, I'm wrapping now the value of `conf.getString` to an `Option` and still use `getOrElse`, mainly because it feels more scala-ish to use `getOrElse` for a default value instead of relying on a custom API to handle default values.